platforms/web/lib/useListeners/useListeners.ts

```diff
@@ -123,15 +123,24 @@ export function useListeners(
         const onInput = (e: Event) => isInputEvent(e) && _handleInput(e);
         editorNode.addEventListener('input', onInput);
 
-        const onPaste = (e: Event) => {
-            if (!isClipboardEvent(e)) {
-                return;
+        // Can be called by onPaste or onBeforeInput
+        const onPaste = (e: ClipboardEvent | InputEvent) => {
+            // this is required to handle edge case image pasting in Safari, see
+            // https://github.com/vector-im/element-web/issues/25327
+            const isSpecialCaseInputEvent =
+                isInputEvent(e) &&
+                e.inputType === 'insertFromPaste' &&
+                e.dataTransfer !== null;
+
+            const isEventToHandle =
+                isClipboardEvent(e) || isSpecialCaseInputEvent;
+
+            if (isEventToHandle) {
+                e.preventDefault();
+                e.stopPropagation();
+
+                _handleInput(e);
```

Comment:
Before we only used to call this from `onPaste` and returned early (ie did nothing) if it was not an event from the clipboard.

Now we need to handle a special case as a clipboard event, so the check has become "check if it is a clipboard event or the special case, and if it's either of those, do stuff. Otherwise, do nothing"